hai/parallel.py

```diff
@@ -238,3 +216,59 @@ def exceptions(self) -> Dict[str, Exception]:
             for t in self.tasks
             if t.ready() and not t._success  # type: ignore[attr-defined]
         }
+
+
+class ChordParallelRun(BaseParallelRun):
+    """Parallel run that inherits the thread pool from another parallel run
+    instead of managing its own
+    """
+
+    def __init__(self, main_run: "ParallelRun") -> None:
+        super().__init__()
+        self.main_run = main_run
+        self.pool = main_run.pool
+
+    def ready(self) -> bool:
+        return all([t.ready() for t in self.tasks])
+
+
+class ParallelRun(BaseParallelRun):
+    """
+    Encapsulates running several functions in parallel.
+
+    Due to the GIL, this is mainly useful for IO-bound threads, such as
+    downloading stuff from the Internet, or writing big buffers of data.
+
+    It's recommended to use this in a `with` block, to ensure the thread pool
+    gets properly shut down.
+    """
+
+    def __init__(self, parallelism: Optional[int] = None) -> None:
+        super().__init__()
+        self.pool = ThreadPool(
+            processes=(parallelism or (int(os.cpu_count() or 1) * 2)),
+        )
+        self.chords: WeakSet[ChordParallelRun] = WeakSet()
+
+    def chord(self) -> ChordParallelRun:
+        """Return a chord ParallelRun that can run a set of tasks using
+        the same thread pool as this run, but have their separate task pool that
+        can be waited separately.
+
+        The chord should be run and waited to completion during the lifetime of this object,
+        as it will terminate the thread pool on deallocation or context manager exit.
```

Comment:
So the chord itself isn't counted as a task for the parent `pr.wait()`? Just so I understand things right.

Reply:
It's not, unless the chord is run and waited for inside a task of the ParallelRun. This feature is intended for, but not necessarily limited to such patterns where tasks being run inside the ParallelRun would like to run additional tasks of their own, and manage their life cycle.

Exiting the main parallel run will still terminate the thread pool and therefore any chords with tasks.

Reply:
Docstring improved